File: 210. Course Schedule II/findOrder.py
from collections import deque
from typing import List

class Solution:
  def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
    graph = { i: [] for i in range(numCourses) } 
    indegree = { i: 0 for i in range(numCourses) }
    
    for dest, src in prerequisites:
      graph[src].append(dest)
      indegree[dest] += 1
      
    queue = deque([i for i in indegree if indegree[i] == 0])
    
    res = []
    while queue:
      node = queue.popleft()
      res.append(node)
      for neighbor in graph[node]:
        indegree[neighbor] -= 1
        if indegree[neighbor] == 0:
          queue.append(neighbor)
          
    return res if len(res) == numCourses else []


numCourses = 4
prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]]
solution = Solution()
print('findOrder: ', solution.findOrder(numCourses, prerequisites))

File: 210. Course Schedule II/findOrder.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
const findOrder = (numCourses, prerequisites) => {
  const graph = {};
  const inDegree = {};

  for (let i = 0; i < numCourses; i++) {
    graph[i] = [];
    inDegree[i] = 0;
  }

  for (let i = 0; i < prerequisites.length; i++) {
    const [dest, src] = prerequisites[i];
    graph[src].push(dest);
    inDegree[dest]++;
  }

  const queue = Object.keys(inDegree).filter(item => inDegree[item] === 0);

  const res = [];
  while (queue.length) {
    const node = queue.shift()
    res.push(parseInt(node));
    const neighbors = graph[node];

    for (let i = 0; i < neighbors.length; i++) {
      inDegree[neighbors[i]]--;
      if (inDegree[neighbors[i]] === 0) {
        queue.push(neighbors[i]);
      }
    }
  }

  return res.length === numCourses ? res : [];
};


const numCourses = 4;
const prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]];
console.log('findOrder: ', findOrder(numCourses, prerequisites)); // [0, 1, 2, 3]
